### PR TITLE
very dumb leader selection

### PR DESCRIPTION
**Pull request by aeyakovenko**
* very dumb leader selection (fixes part 3 of #299)

### CODE REVIEW

File: src/bin/fullnode.rs [hunk 1]
```diff
@@ -36,7 +36,12 @@ fn main() {
     let mut opts = Options::new();
     opts.optflag("h", "help", "print help");
     opts.optopt("l", "", "run with the identity found in FILE", "FILE");
-    opts.optopt("v", "", "validate; find leader's identity in FILE", "FILE");
+    opts.optopt(
+        "t",
+        "",
+        "testnet; connec to the network at this gossip entry point",
```
Comment (garious) [inline src/bin/fullnode.rs]: Connect

File: src/crdt.rs [hunk 2]
```diff
@@ -551,6 +551,33 @@ impl Crdt {
         blob_sender.send(q)?;
         Ok(())
     }
+    /// FIXME: This is obviously the wrong way to do this. Need to implement leader selection
```
Comment (garious) [inline src/crdt.rs]: Let's consistently used TODO or FIXME. Codebase currently uses TODO. If you prefer FIXME, can you change the others?

File: src/bin/fullnode.rs [hunk 3]
```diff
@@ -119,14 +124,14 @@ fn main() {
...
         eprintln!(
-            "starting validator... {} using {}",
-            repl_data.requests_addr, path
+            "starting validator... {} connecting to {}",
+            repl_data.requests_addr, testnet
         );
-        let file = File::open(path.clone()).expect(&format!("file not found: {}", path));
-        let leader = serde_json::from_reader(file).expect("parse");
+        let taddr = testnet.parse().unwrap();
+        let entry = ReplicatedData::new_entry_point(taddr);
```
Comment (garious) [inline src/bin/fullnode.rs]: The term `entry` is already actively used throughout the codebase. Can you find another?

File: src/bin/fullnode.rs [hunk 4]
```diff
@@ -119,14 +124,14 @@ fn main() {
...
+        let testnet = matches.opt_str("t").unwrap();
         eprintln!(
-            "starting validator... {} using {}",
-            repl_data.requests_addr, path
+            "starting validator... {} connecting to {}",
+            repl_data.requests_addr, testnet
         );
-        let file = File::open(path.clone()).expect(&format!("file not found: {}", path));
-        let leader = serde_json::from_reader(file).expect("parse");
+        let taddr = testnet.parse().unwrap();
```
Comment (garious) [inline src/bin/fullnode.rs]: `testnet_addr`?

File: src/crdt.rs [hunk 5]
```diff
@@ -551,6 +551,33 @@ impl Crdt {
...
+            *cnt += 1;
+            println!("leader {:?} {}", &v.current_leader_id[..4], *cnt);
+        }
+        let mut sorted: Vec<_> = table.iter().collect();
+        sorted.sort_by_key(|a| a.1);
+        sorted.last().map(|a| *(*(*a).0))
+    }
+
+    /// FIXME: This is obviously the wrong way to do this. Need to implement leader selection
+    /// A t-shirt for the first person to actually use this bad behavior to attack the alpha testnet
```
Comment (garious) [inline src/crdt.rs]: Seems fair :)

File: src/crdt.rs [hunk 6]
```diff
@@ -551,6 +551,33 @@ impl Crdt {
...
+    fn update_leader(&mut self) {
+        if let Some(lid) = self.top_leader() {
```
Comment (garious) [inline src/crdt.rs]: `id` would be a better name since `lid` is a word. If not specific enough, use `leader_id`

File: src/crdt.rs [hunk 7]
```diff
@@ -1173,4 +1206,29 @@ mod tests {
             assert_eq!(blob.get_id().unwrap(), id);
         }
     }
+    /// FIXME: This is obviously the wrong way to do this. Need to implement leader selection,
+    /// delete this test after leader selection is correctly implemented
+    #[test]
+    fn test_update_leader() {
+        logger::setup();
+        let me = ReplicatedData::new_leader(&"127.0.0.1:1234".parse().unwrap());
+        let lead = ReplicatedData::new_leader(&"127.0.0.1:1234".parse().unwrap());
```
Comment (garious) [inline src/crdt.rs]: `leader0` and `leader1` would be consistent with naming elsewhere

File: src/bin/fullnode.rs [hunk 8]
```diff
@@ -36,7 +36,12 @@ fn main() {
     let mut opts = Options::new();
     opts.optflag("h", "help", "print help");
     opts.optopt("l", "", "run with the identity found in FILE", "FILE");
-    opts.optopt("v", "", "validate; find leader's identity in FILE", "FILE");
+    opts.optopt(
+        "t",
+        "",
+        "testnet; connec to the network at this gossip entry point",
+        "host:port",
```
Comment (rob-solana) [inline src/bin/fullnode.rs]: recommend HOST:PORT

File: src/crdt.rs [hunk 9]
```diff
@@ -1173,4 +1206,29 @@ mod tests {
...
+        let lead2 = ReplicatedData::new_leader(&"127.0.0.1:1234".parse().unwrap());
+        let mut crdt = Crdt::new(me.clone());
+        assert_matches!(crdt.top_leader(), None);
```
Comment (garious) [inline src/crdt.rs]: `assert_eq`

File: src/bin/fullnode.rs (hunk 1 above)
Comment (rob-solana) [inline src/bin/fullnode.rs]: s/connec to/connect to at gossip entry point HOST:PORT

File: src/crdt.rs [hunk 10]
```diff
@@ -577,14 +603,20 @@ impl Crdt {
...
                 if exit.load(Ordering::Relaxed) {
                     return;
                 }
                 //TODO: possibly tune this parameter
                 //we saw a deadlock passing an obj.read().unwrap().timeout into sleep
-                sleep(Duration::from_millis(GOSSIP_SLEEP_MILLIS));
+                let _ = obj.write().unwrap().update_leader();
+                let elapsed = timestamp() - start;
+                if GOSSIP_SLEEP_MILLIS > elapsed {
+                    let left = GOSSIP_SLEEP_MILLIS - elapsed;
```
Comment (garious) [inline src/crdt.rs]: `left` is unnecessarily ambiguous. `time_left` or `remaining` would be better.

File: src/crdt.rs [hunk 11]
```diff
@@ -551,6 +551,32 @@ impl Crdt {
...
+        let def = PublicKey::default();
+        let cur = self.table.values().filter(|x| x.current_leader_id != def);
+        for v in cur {
+            let cnt = table.entry(&v.current_leader_id).or_insert(0);
+            *cnt += 1;
+            trace!("leader {:?} {}", &v.current_leader_id[..4], *cnt);
+        }
+        let mut sorted: Vec<_> = table.iter().collect();
+        sorted.sort_by_key(|a| a.1);
+        sorted.last().map(|a| *(*(*a).0))
```
Comment (garious) [inline src/crdt.rs]: Can that be simplified?

File: src/crdt.rs (hunk 9 above)
Comment (aeyakovenko) [reply src/crdt.rs]: I don't think that will type check, it will complain that None is untyped I think.

File: src/crdt.rs (hunk 2 above)
Comment (aeyakovenko) [reply src/crdt.rs]: I thought FIXME was a higher priority todo :)

File: src/crdt.rs (hunk 11 above)
Comment (aeyakovenko) [reply src/crdt.rs]: hehe, its borrow checker all the way down

Comment (aeyakovenko) [reply src/crdt.rs]: I was wrong

**Comment by aeyakovenko** [comment]
@garious still blocked?

**Comment by garious** [comment]
Merged in #430

**Pull request closed without merging**